### PR TITLE
Grazing angle metalness

### PR DESCRIPTION
**Pull request by bhouston**
This is built upon the simplified lighting PR: #7324

This fixes an issue with Physical Material in that the grazing angles of the materials with high metalness should use the metal color, rather than going to white (mentioned in #7397).  This is because metals should be conductive fresnel (goes to metal color at grazing angles) rather than dielectric fresnel (goes to white at grazing angles).  I implemented this via allowing F_Shlick to take a grazingColor.  One could alternatively implement this via a mix of F_Shchlick result with the diffuse color based on metalness, but that was more awkward to implement, because it would make the GGX Specular BRDF aware internally of metalness.

### CODE REVIEW

**Comment by mrdoob** [comment]
@WestLangley thoughts?

**Comment by WestLangley** [comment]
Let's discuss the implementation details after things settle down. In principle, it sounds reasonable.

**Comment by bhouston** [comment]
This PR has been updated to work with the changes in dev.

**Comment by mrdoob** [comment]
@WestLangley looks good?

**Comment by WestLangley** [comment]
There is something I do not understand...

> This is because metals should be conductive fresnel (goes to metal color at grazing angles) rather than dielectric fresnel (goes to white at grazing angles)

Have a look at figure 5 in [Schilck's original paper](http://igorsklyar.com/system/documents/papers/28/Schlick94.pdf). He seems to imply otherwise.

Also [here](https://disney-animation.s3.amazonaws.com/library/s2012_pbs_disney_brdf_notes_v2.pdf) in section 5.5.

There are other models for air-to-conductor Fresnel. See for example https://seblagarde.wordpress.com/2013/04/29/memo-on-fresnel-equations/? and http://sirkan.iit.bme.hu/~szirmay/fresnel.pdf

File: src/renderers/shaders/ShaderChunk/bsdfs.glsl [hunk 1]
```diff
@@ -28,7 +28,7 @@ vec3 F_Schlick( const in vec3 specularColor, const in float dotLH ) {
 	// Optimized variant (presented by Epic at SIGGRAPH '13)
 	float fresnel = exp2( ( -5.55437 * dotLH - 6.98316 ) * dotLH );
 
-	return ( 1.0 - specularColor ) * fresnel + specularColor;
+	return mix( specularColor, grazingColor, fresnel );
```
Comment (WestLangley) [inline src/renderers/shaders/ShaderChunk/bsdfs.glsl]: With this change, this is no longer Schlick's infamous model, so the method name would need to be changed. `F_SchlickHouston`?

File: src/renderers/shaders/ShaderChunk/bsdfs.glsl [hunk 2]
```diff
@@ -107,7 +107,7 @@ vec3 BRDF_Specular_GGX_Environment( const in GeometricContext geometry, vec3 spe
 
 	vec2 AB = vec2( -1.04, 1.04 ) * a004 + r.zw;
 
-	return specularColor * AB.x + AB.y;
+	return ( specularColor * AB.x + AB.y ) * grazingColor;
```
Comment (WestLangley) [inline src/renderers/shaders/ShaderChunk/bsdfs.glsl]: I am not sure this is the right modification here...

File: src/renderers/shaders/ShaderChunk/bsdfs.glsl (hunk 1 above)
Comment (bhouston) [inline src/renderers/shaders/ShaderChunk/bsdfs.glsl]: It is equivalent to F_Schlick in the case that the grazingColor is white.

File: src/renderers/shaders/ShaderChunk/bsdfs.glsl (hunk 2 above)
Comment (bhouston) [inline src/renderers/shaders/ShaderChunk/bsdfs.glsl]: We do need to have an unoptimized version of the environment BRDF.  This formulation is correct at the bounds of the metalness function, whether it is correct within the middle is harder to known as I am not sure the answer is clear.

File: src/renderers/shaders/ShaderChunk/bsdfs.glsl (hunk 1 above)
Comment (bhouston) [inline src/renderers/shaders/ShaderChunk/bsdfs.glsl]: Remember that F_Schlick is designed as a dielectric Fresnel function that is not valid for metal objects that exhibit conductor Fresnel reflection.  Conductor Fresnel does not necessarily go to white at grazing angles, rather it a specific color defined by the properties of the metal -- a yellowish color for things like gold.

We could calculate a separate conductor Fresnel and blend between it and the F_Schlick dieletric Fresnel?  I think that a fast conductor Fresnel approximation is likely nearly what I created.  The only real enhancement one could make would be to allow for direct specification of the grazingColor as it does differ a bit from the F0 color.

**Comment by bhouston** [comment]
Oh, according to your link it does go to white, just much later than dielectric:

https://seblagarde.wordpress.com/2013/04/29/memo-on-fresnel-equations/

**Comment by bhouston** [comment]
It is interesting that there is an error term for F_Schick to compensate for the metal difference:

```
Lazanyi et al [8] try to provide a better FresnelSchlick approximation for Metals. They try to add a compensation term -a \cos \theta (1 - \cos \theta)^\alpha
```

I would be okay using that instead.

For the Environmental BRDF, I am sure there is a F_Schlick term in there.  IF we could get an non-optimized version of it, we could reuse the same F_SchlickLazanyi term in both direct and indirect specular to model conductor Fresnel more accurately.

**Comment by bhouston** [comment]
I look at this paper's pictures of conductor Fresnel materials as examples of what happens at the edges.  If you look very very closely you can sort of see it going to white, but it does so only at the most perpendicular angles:

http://jcgt.org/published/0003/04/03/paper.pdf

This is why my approximation of just having it go the grazing angle color made sense to me.  But I am fine with a compensation factor based on F_SchlickLazanyi.  Metals really do not look right when they go to white early.

**Comment by WestLangley** [comment]
> Oh, according to your link it does go to white, just much later than dielectric:
> Metals really do not look right when they go to white early.

Let's use one of the published models, such as Gulbrandsen's. What you have proposed in this PR does not have the correct behavior.

**Comment by WestLangley** [comment]
And yes, I agree, we need to understand EnvironmentBRDF. That's on the todo-list.

**Comment by bhouston** [comment]
@WestLangley 
> Let's use one of the published models, such as Gulbrandsen's. What you have proposed in this PR does not have the correct behavior.

This is what I have implemented for my MeshMetalMaterial that I would like to submit once we get through all these PRs.  Although it requires a new public material parameter, an edge tint.  That would be an addition to the Disney PBR model, as it doesn't include it.

I'll close this PR because we do not have a way to modify the opaque Environment BRDF, thus I view that nothing can get accepted until issue is addressed.

**Pull request closed without merging**